### PR TITLE
SDK Support for CDH5.10 and above

### PR DESCRIPTION
**Pull request by aadarshjajodia**

### CODE REVIEW

**Comment by aadarshjajodia** [comment]
@changhiskhan

**Comment by aadarshjajodia** [comment]
@changhiskhan

File: client/pom.xml [hunk 1]
```diff
@@ -20,7 +20,7 @@
   <parent>
     <artifactId>navigator-sdk</artifactId>
     <groupId>com.cloudera.navigator</groupId>
-    <version>1.2-SNAPSHOT</version>
+    <version>2.0</version>
```
Comment (changhiskhan) [inline client/pom.xml]: probably 2.0-SNAPSHOT

File: model/src/test/java/com/cloudera/nav/sdk/model/entities/EntityTest.java [hunk 2]
```diff
@@ -1,72 +0,0 @@
...
-
-import static org.junit.Assert.assertEquals;
-
-import com.cloudera.nav.sdk.model.MD5IdGenerator;
-import com.cloudera.nav.sdk.model.Source;
-import com.cloudera.nav.sdk.model.SourceType;
-
-import org.junit.Test;
-
-public class EntityTest {
```
Comment (changhiskhan) [inline model/src/test/java/com/cloudera/nav/sdk/model/entities/EntityTest.java]: We're removing this unit test? why?

File: model/src/main/java/com/cloudera/nav/sdk/model/entities/Entity.java [hunk 3]
```diff
@@ -344,4 +351,10 @@ public boolean equals(Object o) {
   public int hashCode() {
     return identity.hashCode();
   }
+
+  public Boolean getIsEndPoint() { return isEndPoint; }
+
+  public void setIsEndPoint(Boolean isEndPoint) {
```
Comment (changhiskhan) [inline model/src/main/java/com/cloudera/nav/sdk/model/entities/Entity.java]: So now a users has to know to call this explicitly? It doesn't feel right.

**Comment by changhiskhan** [comment]
Have more review comments. Also, but just wanted to start the conversation early while i'm finishing the review.

File: model/src/test/java/com/cloudera/nav/sdk/model/entities/EntityTest.java (hunk 2 above)
Comment (aadarshjajodia) [inline model/src/test/java/com/cloudera/nav/sdk/model/entities/EntityTest.java]: This Unit test was there to test the ID generation logic in the SDK. We don't have the id generation logic anymore in the SDK

File: model/src/main/java/com/cloudera/nav/sdk/model/entities/Entity.java (hunk 3 above)
Comment (aadarshjajodia) [inline model/src/main/java/com/cloudera/nav/sdk/model/entities/Entity.java]: It is similar to the user having to create an EndPointProxy. Instead of the user having to create that EndPointProxy, the user can simply create the type of Entity they want to link to and set this flag.

File: model/src/test/java/com/cloudera/nav/sdk/model/entities/EntityTest.java (hunk 2 above)
Comment (changhiskhan) [inline model/src/test/java/com/cloudera/nav/sdk/model/entities/EntityTest.java]: Right, so should we be unit testing the new constructors, making sure you must provide either id attrs or id, etc?

File: model/src/main/java/com/cloudera/nav/sdk/model/entities/Entity.java (hunk 3 above)
Comment (changhiskhan) [inline model/src/main/java/com/cloudera/nav/sdk/model/entities/Entity.java]: ok, I thought about why this pattern felt weird. What you're doing is adding a transient attribute to control a completely different level of abstraction that really has nothing to do with the Entity itself. Now the Entity not only has to know about itself, it must also contain knowledge about how the SDK client will convert data during transport.

File: model/src/main/java/com/cloudera/nav/sdk/model/IdAttrs.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+import java.util.Collection;
+import java.util.Map;
+
+import org.apache.commons.lang.StringUtils;
+import org.joda.time.Instant;
+
+/**
+ * Used for deserializing custom entities. We use this as an intermediary
+ * so that we can handle the changesets for tags/UDPs
```
Comment (changhiskhan) [inline model/src/main/java/com/cloudera/nav/sdk/model/IdAttrs.java]: Can you explain the interaction with tags/UDP changesets?

File: model/src/main/java/com/cloudera/nav/sdk/model/IdAttrs.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+ */
+public class IdAttrs {
+
+  @MProperty
```
Comment (changhiskhan) [inline model/src/main/java/com/cloudera/nav/sdk/model/IdAttrs.java]: This isn't an @MClass. Do you need @MProperty annotations here?

File: model/src/main/java/com/cloudera/nav/sdk/model/IdAttrs.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+import java.util.Map;
+
+import org.apache.commons.lang.StringUtils;
+import org.joda.time.Instant;
+
+/**
+ * Used for deserializing custom entities. We use this as an intermediary
+ * so that we can handle the changesets for tags/UDPs
+ */
+public class IdAttrs {
```
Comment (changhiskhan) [inline model/src/main/java/com/cloudera/nav/sdk/model/IdAttrs.java]: Looking at the code, users are not explicitly interacting with IdAttrs. What if we just had a Map instead of a catch-all class?

File: examples/src/main/java/com/cloudera/nav/sdk/examples/extraction/FilteredExtraction.java [hunk 7]
```diff
@@ -61,8 +61,8 @@ public static void main(String[] args) throws IOException {
     MetadataExtractor extractor = new MetadataExtractor(client, null);
 
     // Run filtered examples
-    getHDFSEntities(client, extractor, marker);
-    getHive(extractor, marker, "city_id");
+    //getHDFSEntities(client, extractor, marker);
```
Comment (changhiskhan) [inline examples/src/main/java/com/cloudera/nav/sdk/examples/extraction/FilteredExtraction.java]: Are you disabling this example?

File: client/src/main/java/com/cloudera/nav/sdk/client/writer/registry/RelationFactory.java [hunk 8]
```diff
@@ -106,16 +113,32 @@ private Relation createInstanceOfRelation(RelationRole roleOfOther,
...
         .namespace(namespace).build();
   }
 
+  private Collection<IdAttrs> getIdAttrsList(
+      Collection<? extends Entity> other) {
+    Collection<IdAttrs> idAttrsList = Lists.newArrayList();
+    for (Entity en : other) {
+      if (Strings.isNullOrEmpty(en.getIdentity())) {
+        IdAttrs at = new IdAttrs();
+        en.populateIdAttrs(at);
```
Comment (changhiskhan) [inline client/src/main/java/com/cloudera/nav/sdk/client/writer/registry/RelationFactory.java]: Why not just make a static method so you can combine the constructor call and the population into a single call?

File: model/src/main/java/com/cloudera/nav/sdk/model/relations/Relation.java [hunk 9]
```diff
@@ -178,11 +195,11 @@ public T ep2(Collection<Entity> ep2Entities) {
   private String identity;
   @MProperty(required=true)
   private RelationType type;
-  @MProperty(required=true, attribute="endpoint1Ids")
```
Comment (changhiskhan) [inline model/src/main/java/com/cloudera/nav/sdk/model/relations/Relation.java]: How will you validate relations now?

File: model/src/main/java/com/cloudera/nav/sdk/model/relations/Relation.java [hunk 10]
```diff
@@ -197,8 +214,15 @@ public T ep2(Collection<Entity> ep2Entities) {
...
+  @MProperty
+  private Collection<IdAttrs> ep1Attributes;
+
+  @MProperty
+  private Collection<IdAttrs> ep2Attributes;
+
+
   protected Relation(Builder<?> builder) {
-    Preconditions.checkState(builder.identity != null ||
+    /*Preconditions.checkState(builder.identity != null ||
```
Comment (changhiskhan) [inline model/src/main/java/com/cloudera/nav/sdk/model/relations/Relation.java]: Don't leave code that's commented in.

File: model/src/main/java/com/cloudera/nav/sdk/model/IdAttrs.java (hunk 4 above)
Comment (aadarshjajodia) [inline model/src/main/java/com/cloudera/nav/sdk/model/IdAttrs.java]: Need to update the comment in the file

File: model/src/main/java/com/cloudera/nav/sdk/model/IdAttrs.java (hunk 5 above)
Comment (aadarshjajodia) [inline model/src/main/java/com/cloudera/nav/sdk/model/IdAttrs.java]: @MProperty is added to these fields to make sure that the IdAttrs are written to the JSON

File: model/src/main/java/com/cloudera/nav/sdk/model/relations/Relation.java (hunk 9 above)
Comment (aadarshjajodia) [inline model/src/main/java/com/cloudera/nav/sdk/model/relations/Relation.java]: Will need to add a validation saying that either the ep1Id or ep1Attribtues, one of these must be present for the Relation to be valid.

File: model/src/main/java/com/cloudera/nav/sdk/model/relations/Relation.java (hunk 10 above)
Comment (aadarshjajodia) [inline model/src/main/java/com/cloudera/nav/sdk/model/relations/Relation.java]: Will update, removing the commented code.

File: examples/src/main/java/com/cloudera/nav/sdk/examples/extraction/FilteredExtraction.java (hunk 7 above)
Comment (aadarshjajodia) [inline examples/src/main/java/com/cloudera/nav/sdk/examples/extraction/FilteredExtraction.java]: Fixed. Some sources were returning null as the sourceExtractionIterationId and MetadataExtractor was not being able to handle that. Have added a check for that in MetadataExtractor

File: model/src/main/java/com/cloudera/nav/sdk/model/relations/Relation.java (hunk 9 above)
Comment (aadarshjajodia) [inline model/src/main/java/com/cloudera/nav/sdk/model/relations/Relation.java]: Have added a **validateRelation** method in RelationValidator which ensures that a Relation either the id or the attributes needed to generate the id.

File: client/pom.xml (hunk 1 above)
Comment (aadarshjajodia) [inline client/pom.xml]: changed to 2.0-SNAPSHOT in all the pom files

Comment (aadarshjajodia) [inline client/pom.xml]: I have run all the examples except the HIVE example. The adding tags to Hive entities since we are planning to support other entity types in the next version

File: model/src/main/java/com/cloudera/nav/sdk/model/IdAttrs.java (hunk 6 above)
Comment (aadarshjajodia) [inline model/src/main/java/com/cloudera/nav/sdk/model/IdAttrs.java]: if i understand correctly you mean putting a Map in the Entity object instead of having an Object here? We essentially want to keep this structure because this is what an ep1/ep2 attributes looks like for the Relation object. And also it makes it easy for users to just say hdfsDir.setFileSystemPath()

File: model/src/test/java/com/cloudera/nav/sdk/model/entities/EntityTest.java (hunk 2 above)
Comment (aadarshjajodia) [inline model/src/test/java/com/cloudera/nav/sdk/model/entities/EntityTest.java]: Ok I am planning to add a validateEntity abstract method in Entity class, so that each entity can have its own implementation of the method. Also when we write the entity to the JSON. we can simply call entity.validateEntity() to make sure the required arguments are passed in.

File: client/src/main/java/com/cloudera/nav/sdk/client/MetadataExtractor.java [hunk 11]
```diff
@@ -142,8 +142,10 @@ public MetadataResultSet extractMetadata(String startMarkerRep,
     HashMap<String, Long> newMarker = Maps. newHashMap();
     for (Source source : sources) {
       String id = source.getIdentity();
-      Long sourceExtractIteration = (current) ?
```
Comment (changhiskhan) [inline client/src/main/java/com/cloudera/nav/sdk/client/MetadataExtractor.java]: why this change?

File: model/src/main/java/com/cloudera/nav/sdk/model/entities/Entity.java (hunk 3 above)
Comment (changhiskhan) [inline model/src/main/java/com/cloudera/nav/sdk/model/entities/Entity.java]: I feel pretty strongly we shouldn't do this.

Comment (changhiskhan) [inline model/src/main/java/com/cloudera/nav/sdk/model/entities/Entity.java]: Hopefully the reasons are clear. If not please let me know.

File: model/src/main/java/com/cloudera/nav/sdk/model/entities/HiveColumn.java [hunk 12]
```diff
@@ -91,4 +91,6 @@ public void setColumnName(String columnName) {
     setName(columnName);
   }
 
+  @Override
+  public void validateEntity() {}
```
Comment (changhiskhan) [inline model/src/main/java/com/cloudera/nav/sdk/model/entities/HiveColumn.java]: No validation needed?

File: model/src/main/java/com/cloudera/nav/sdk/model/entities/PigOperation.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    this();
+    setIdentity(id);
+  }
+
+  @Override
+  public void validateEntity() {
+    if (Strings.isNullOrEmpty(this.getIdentity()) &&
+        (Strings.isNullOrEmpty(this.getJobName()) ||
+        Strings.isNullOrEmpty(this.getLogicalPlanHash())))
+      throw new IllegalArgumentException(
```
Comment (changhiskhan) [inline model/src/main/java/com/cloudera/nav/sdk/model/entities/PigOperation.java]: Use Preconditions here?

File: model/src/main/java/com/cloudera/nav/sdk/model/relations/RelationIdGenerator.java [hunk 14]
```diff
@@ -15,6 +15,7 @@
  */
 package com.cloudera.nav.sdk.model.relations;
 
+import com.cloudera.nav.sdk.model.IdAttrs;
```
Comment (changhiskhan) [inline model/src/main/java/com/cloudera/nav/sdk/model/relations/RelationIdGenerator.java]: unused?

File: model/src/main/java/com/cloudera/nav/sdk/model/relations/RelationValidator.java [hunk 15]
```diff
@@ -61,4 +61,15 @@ public void validateRequiredMProperties(Relation relation) {
       }
     }
   }
+
+  public void validatateRelation(Relation relation) {
+    Preconditions.checkArgument(
```
Comment (changhiskhan) [inline model/src/main/java/com/cloudera/nav/sdk/model/relations/RelationValidator.java]: Are there other relations validation criteria other than these? If you want to add additional single field validation criteria you'd have to explicitly check for it now?

File: model/src/main/java/com/cloudera/nav/sdk/model/IdAttrs.java [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+import java.util.Map;
+
+import org.apache.commons.lang.StringUtils;
+import org.joda.time.Instant;
+
+/**
+ * Used for specifying the attributes needed for generation of entity id's at
+ * the server side.
+ */
+public class IdAttrs {
```
Comment (changhiskhan) [inline model/src/main/java/com/cloudera/nav/sdk/model/IdAttrs.java]: Can we revisit doing this on the client side? The more I look at this the more it seems too dirty from a design perspective. There must be a better way than essentially blowing up the abstractions we've built around the different Entity classes.

File: examples/src/main/java/com/cloudera/nav/sdk/examples/tags/SetHdfsFileTags.java [hunk 17]
```diff
@@ -45,10 +46,11 @@ public static void main(String[] args) {
     Source fs = client.getSourcesForType(SourceType.HDFS).iterator().next();
 
     // send tags for multiple entities to Navigator
-    HdfsEntity dir = new HdfsEntity("/user/hdfs", EntityType.DIRECTORY,
+    HdfsEntity dir = new HdfsEntity("/user/hdfs", EntityType
+        .DIRECTORY,
         fs.getIdentity());
-    dir.setTags(Sets.newHashSet("HAS_SENSITIVE_FILES",
-        "CONTAINS_SOME_SUPER_SECRET_STUFF"));
+    dir.setTags(Sets.newHashSet("TAG_1",
```
Comment (changhiskhan) [inline examples/src/main/java/com/cloudera/nav/sdk/examples/tags/SetHdfsFileTags.java]: You don't like my tag names ? :smile:

File: examples/src/main/java/com/cloudera/nav/sdk/examples/schema/FireCircleDataset.java [hunk 18]
```diff
@@ -61,5 +62,9 @@ public void setDataContainer(HdfsEntity hdfsDir) {
     Preconditions.checkArgument(hdfsDir.getEntityType() ==
         EntityType.DIRECTORY);
     this.dataContainer = hdfsDir;
+    this.dataContainer.setIsEndPoint(true);
   }
+
+  @Override
+  public void validateEntity() {}
```
Comment (changhiskhan) [inline examples/src/main/java/com/cloudera/nav/sdk/examples/schema/FireCircleDataset.java]: If you're having to create empty stubs all around your code like this, you're probably not doing the right thing

File: examples/src/main/java/com/cloudera/nav/sdk/examples/tags/SetHdfsFileTags.java (hunk 17 above)
Comment (aadarshjajodia) [inline examples/src/main/java/com/cloudera/nav/sdk/examples/tags/SetHdfsFileTags.java]: not really. There is actually no change needed in this file. Will update

File: model/src/main/java/com/cloudera/nav/sdk/model/entities/Entity.java (hunk 3 above)
Comment (aadarshjajodia) [inline model/src/main/java/com/cloudera/nav/sdk/model/entities/Entity.java]: sure. Lets discuss about this tomorrow. We need a way to differentiate when to write certain entities to the json and when to not write them. The EndPointProxy achieved this same purpose.

**Comment by aadarshjajodia** [comment]
Hi @changhiskhan, as discussed I have removed the IdAttrs. Now each entity will return an IdAttrsMap which can be set directly to the ep1/ep2 attributes in a Relation. Also i have made the method validateEntity to be non-abstract and removed empty stubs.

File: client/src/main/java/com/cloudera/nav/sdk/client/MetadataExtractor.java (hunk 11 above)
Comment (changhiskhan) [reply client/src/main/java/com/cloudera/nav/sdk/client/MetadataExtractor.java]: ?

File: model/src/test/java/com/cloudera/nav/sdk/model/entities/EntityTest.java [hunk 19]
```diff
@@ -1,72 +1,54 @@
...
-        entity.getFileSystemPath());
-    assertEquals(id, entity.generateId());
+  @Test(expected = IllegalArgumentException.class)
+  public void testInvalidPigOperation() {
+    PigOperation pigOperation = new PigOperation();
+    pigOperation.validateEntity();
   }
 
   @Test
-  public void testHiveTable() {
```
[the remark this answers is not in the document]
Comment (aadarshjajodia) [reply model/src/test/java/com/cloudera/nav/sdk/model/entities/EntityTest.java]: We are not supporting Hive in the first version of this new sdk release. Will be adding it back in the next version

File: model/src/main/java/com/cloudera/nav/sdk/model/entities/PigOperation.java (hunk 13 above)
Comment (aadarshjajodia) [reply model/src/main/java/com/cloudera/nav/sdk/model/entities/PigOperation.java]: Preconditions checks for an argument to be true. Something like Preconditions.checkArgument() and it throws the error if the expression is false. We can put Preconditions here but it will just add more of !, which will make these conditions look messy.

File: model/src/main/java/com/cloudera/nav/sdk/model/relations/RelationIdGenerator.java (hunk 14 above)
Comment (aadarshjajodia) [reply model/src/main/java/com/cloudera/nav/sdk/model/relations/RelationIdGenerator.java]: Removed already

File: model/src/main/java/com/cloudera/nav/sdk/model/IdAttrs.java (hunk 16 above)
Comment (aadarshjajodia) [reply model/src/main/java/com/cloudera/nav/sdk/model/IdAttrs.java]: have removed this class altogether since we don't need this anymore

File: client/src/main/java/com/cloudera/nav/sdk/client/writer/MetadataWriter.java [hunk 20]
```diff
@@ -125,7 +125,7 @@ private void getMRelations(Entity entity, MClassWrapper graph) {
         entity.getClass());
     for (MRelationEntry relEntry : relationAttrs) {
       for(Entity other : relEntry.getConnectedEntities(entity)) {
-        if (!(other instanceof EndPointProxy)) {
+        if (!other.getIsEndPoint()) {
```
[the remark this answers is not in the document]
Comment (aadarshjajodia) [reply client/src/main/java/com/cloudera/nav/sdk/client/writer/MetadataWriter.java]: Ok. Lets look at this from a user perspective. So lets take our three use cases.
1. User wants to Pre Register an HdfsEntity.
2. User wants to add tags/ custom metadata to an existing HdfsEntity.
3. User wants to link to a already existing HdfsEntity. 

In all these cases user should be just able to directly use the HdfsEntity object that we have in our models. I guess it not make a good design if we had HdfsEndPointEntity and HdfsEntity as separate classes to deal with this. Also if we just make HdfsEntity extend from an EndPointProxy then it restricts that HdfsEntity to be used for the first 2 use cases.

Lets say we have an EndPointProxy class, which can just have the Map of idAttributes. Now when the user is linking to a PigOperation, the user creates an EndPointProxy object. But how does the user know what are the attributes to be passed in to this map of idAttributes? It would have been much easier for the user to just create a PigOperation and set the end point proxy flag. The validate method of PigOperation would validate that is has all the entity attributes required to populate the id for the relation or the entity id. Let me know your thoughts on this.

**Comment by changhiskhan** [comment]
Also, do you have a solution for the duplicate relations problem? If not, can you ask Vikas for help?

File: client/src/main/java/com/cloudera/nav/sdk/client/writer/MetadataWriter.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (aadarshjajodia) [reply client/src/main/java/com/cloudera/nav/sdk/client/writer/MetadataWriter.java]: 1. Custom Lineage example
Since entity x is set as an end point, this hdfs entity will be treated as the end point of the relation. Also since x is not set an end point, it is not written in the elements json.

2. Since a is not set as an End Point it will be written to the elements json.

File: client/src/main/java/com/cloudera/nav/sdk/client/writer/registry/MRelationEntry.java [hunk 21]
```diff
@@ -65,6 +65,7 @@ public Relation buildRelation(Entity entity, String namespace) {
     RelationType type = getRelationTypeFromRole(roleOfOther);
     RelationFactory factory = new RelationFactory();
     Collection<? extends Entity> otherEndpoints = getConnectedEntities(entity);
+
```
[the remark this answers is not in the document]
Comment (aadarshjajodia) [reply client/src/main/java/com/cloudera/nav/sdk/client/writer/registry/MRelationEntry.java]: Removed.

File: model/src/test/java/com/cloudera/nav/sdk/model/PigIdGeneratorTest.java [hunk 22]
```diff
@@ -1,50 +0,0 @@
...
-import static org.junit.Assert.assertEquals;
-
-import com.google.common.collect.Maps;
-
-import java.util.Map;
-
-import org.apache.commons.configuration.MapConfiguration;
-import org.junit.Before;
-
-public class PigIdGeneratorTest {
```
[the remark this answers is not in the document]
Comment (aadarshjajodia) [reply model/src/test/java/com/cloudera/nav/sdk/model/PigIdGeneratorTest.java]: We are not supporting hive in the first version and hence no Hive tests. For Pig, yes i will be adding tests for PigOperation and PigOperationExecution to test the id attrs.

File: model/src/main/java/com/cloudera/nav/sdk/model/relations/Relation.java [hunk 23]
```diff
@@ -197,17 +214,15 @@ public T ep2(Collection<Entity> ep2Entities) {
...
-        builder.idGenerator != null);
-    if (builder.identity != null) {
-      this.identity = builder.identity;
-    } else {
-      this.identity = builder.idGenerator.generateRelationIdentity(
-          builder.ep1Ids, builder.ep1SourceType,
-          builder.ep2Ids, builder.ep2SourceType, builder.type,
-          builder.namespace);
-    }
+    this.identity = UUID.randomUUID().toString();
```
[the remark this answers is not in the document]
Comment (aadarshjajodia) [reply model/src/main/java/com/cloudera/nav/sdk/model/relations/Relation.java]: This identity field is only for the purpose of uniquely identifying each Relation while writing to the relations json to be sent to the server

File: model/src/main/java/com/cloudera/nav/sdk/model/entities/HdfsEntity.java [hunk 24]
```diff
@@ -42,37 +41,33 @@ public HdfsEntity(String fileSystemPath, EntityType type, String sourceId) {
...
 
-  /**
-   * @return the full path for this file or directory
-   */
-  public String getFileSystemPath() {
-    return fileSystemPath;
+  @Override
+  public void validateEntity() {
+    if (Strings.isNullOrEmpty(this.getIdentity()) &&
+        Strings.isNullOrEmpty(this.getFileSystemPath())) {
```
[the remark this answers is not in the document]
Comment (aadarshjajodia) [reply model/src/main/java/com/cloudera/nav/sdk/model/entities/HdfsEntity.java]: Done.

File: examples/src/main/java/com/cloudera/nav/sdk/examples/lineage/StetsonScript.java [hunk 25]
```diff
@@ -49,8 +51,13 @@ public StetsonScript(String namespace) {
    */
   @Override
   public String generateId() {
-    return CustomIdGenerator.generateIdentity(getNamespace(),
-        getPigOperation().getIdentity());
+    if (!Strings.isNullOrEmpty(getPigOperation().getIdentity())) {
```
[the remark this answers is not in the document]
Comment (aadarshjajodia) [reply examples/src/main/java/com/cloudera/nav/sdk/examples/lineage/StetsonScript.java]: Done. The string id of a Stetson Script would be generated using a combination of the namespace and the name of the Stetson Script

File: model/src/main/java/com/cloudera/nav/sdk/model/entities/Entity.java [hunk 26]
```diff
@@ -344,4 +351,16 @@ public boolean equals(Object o) {
...
+    this.isEndPoint = isEndPoint;
+  }
+
+  public Map<String, String> getIdAttrsMap() {
```
[the remark this answers is not in the document]
Comment (aadarshjajodia) [reply model/src/main/java/com/cloudera/nav/sdk/model/entities/Entity.java]: Not every entity has to do this. Only Hadoop Entities need to implement this method. Custom Entities don't need. Hence this is not made abstract.

File: model/src/main/java/com/cloudera/nav/sdk/model/entities/HdfsEntity.java [hunk 27]
```diff
@@ -42,37 +41,33 @@ public HdfsEntity(String fileSystemPath, EntityType type, String sourceId) {
...
     this.fileSystemPath = fileSystemPath;
   }
+
+  public String getFileSystemPath() {
+    return this.fileSystemPath;
+  }
+
+  @Override
+  public Map<String, String> getIdAttrsMap() {
+    return ImmutableMap.of("fileSystemPath", this.getFileSystemPath());
```
[the remark this answers is not in the document]
Comment (aadarshjajodia) [reply model/src/main/java/com/cloudera/nav/sdk/model/entities/HdfsEntity.java]: Here we are only talking only the attributes specific to each entity. When the relation object is created, it populated the source id for respective end point.

File: model/src/main/java/com/cloudera/nav/sdk/model/entities/PigOperation.java [hunk 28]
```diff
@@ -0,0 +1,66 @@
+package com.cloudera.nav.sdk.model.entities;
```
[the remark this answers is not in the document]
Comment (aadarshjajodia) [reply model/src/main/java/com/cloudera/nav/sdk/model/entities/PigOperation.java]: done

File: client/src/test/java/com/cloudera/nav/sdk/client/JsonMetadataWriterTest.java [hunk 29]
```diff
@@ -194,12 +210,8 @@ public void testWriteRelation() throws IOException {
...
     assertEquals(values.get("endpoint2Type"), EntityType.DIRECTORY.name());
     assertTrue(Boolean.valueOf(values.get("userSpecified").toString()));
-    assertEquals(Iterables.getOnlyElement(
-            (Collection<String>) values.get("endpoint1Ids")),
-        inputData.getIdentity());
-    assertEquals(Iterables.getOnlyElement(
-            (Collection<String>) values.get("endpoint2Ids")),
-        outputData.getIdentity());
+
+    //TODO Need to assert the values of ep1Attributes and ep2Attributes
```
[the remark this answers is not in the document]
Comment (aadarshjajodia) [reply client/src/test/java/com/cloudera/nav/sdk/client/JsonMetadataWriterTest.java]: Done

File: client/src/main/java/com/cloudera/nav/sdk/client/writer/registry/RelationFactory.java [hunk 30]
```diff
@@ -59,10 +62,12 @@ private Relation createDataFlowRelation(RelationRole roleOfOther,
                                           Collection<? extends Entity> other,
                                           String namespace) {
     DataFlowRelation.Builder builder = DataFlowRelation.builder();
+    Collection<Map<String, String>> idAttrsList = getIdAttrsList(other);
+
     if (roleOfOther == RelationRole.SOURCE) {
-      builder.target(entity).sources(other);
+      builder.target(entity).sources(other).ep1Attributes(idAttrsList);
```
[the remark this answers is not in the document]
Comment (aadarshjajodia) [reply client/src/main/java/com/cloudera/nav/sdk/client/writer/registry/RelationFactory.java]: There aren't any tests at all for RelationFactory. We can add a unit test for this entire class later.

[the remark this answers is not in the document]
Comment (aadarshjajodia) [reply client/src/main/java/com/cloudera/nav/sdk/client/writer/registry/RelationFactory.java]: There aren't any tests at all for RelationFactory. We can add a unit test for this entire class later.

File: client/src/main/java/com/cloudera/nav/sdk/client/MetadataExtractor.java [hunk 31]
```diff
@@ -140,12 +140,10 @@ public MetadataResultSet extractMetadata(String startMarkerRep,
...
     Collection<Source> sources = client.getAllSources();
     HashMap<String, Long> newMarker = Maps. newHashMap();
+
     for (Source source : sources) {
       String id = source.getIdentity();
-      Long sourceExtractIteration = source.getSourceExtractIteration();
-      if (sourceExtractIteration == null) {
-        sourceExtractIteration = 0L;
-      }
+      Long sourceExtractIteration = current ? source.getSourceExtractIteration() : 0L;
```
Comment (changhiskhan) [inline client/src/main/java/com/cloudera/nav/sdk/client/MetadataExtractor.java]: hmmm so what happens now if getSourceExtractIteration() returns null?

File: examples/src/main/java/com/cloudera/nav/sdk/examples/lineage/StetsonExecution.java [hunk 32]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2015 Cloudera, Inc.
+ * Copyright (c) 2017 Cloudera, Inc.
```
Comment (changhiskhan) [inline examples/src/main/java/com/cloudera/nav/sdk/examples/lineage/StetsonExecution.java]: This doesn't need to change.

File: examples/src/main/java/com/cloudera/nav/sdk/examples/managed/StetsonScript.java [hunk 33]
```diff
@@ -51,13 +48,7 @@ public StetsonScript(String namespace) {
...
   @Override
   public String generateId() {
-    if (!Strings.isNullOrEmpty(getPigOperation().getIdentity())) {
-      return CustomIdGenerator.generateIdentity(getNamespace(),
-          getPigOperation().getIdentity());
-    } else {
-      return CustomIdGenerator.generateIdentity(getNamespace(),
-          UUID.randomUUID().toString());
-    }
+    return CustomIdGenerator.generateIdentity(getNamespace(), getName());
```
Comment (changhiskhan) [inline examples/src/main/java/com/cloudera/nav/sdk/examples/managed/StetsonScript.java]: We should add some comments here and in other custom classes that these fields must uniquely identify the entity

File: model/src/main/java/com/cloudera/nav/sdk/model/entities/HdfsEntity.java [hunk 34]
```diff
@@ -48,13 +50,19 @@ public HdfsEntity(String id) {
...
   public void validateEntity() {
-    if (Strings.isNullOrEmpty(this.getIdentity()) &&
-        Strings.isNullOrEmpty(this.getFileSystemPath())) {
+    if ((Strings.isNullOrEmpty(this.getIdentity()) &&
+        Strings.isNullOrEmpty(this.getFileSystemPath())) ||
+        Strings.isNullOrEmpty(this.getSourceId())) {
       throw new IllegalArgumentException(
           "Either the Entity Id or file system path used" +
-              " to generate the id must be present");
+              " to generate the id must be present along with the source id");
```
Comment (changhiskhan) [inline model/src/main/java/com/cloudera/nav/sdk/model/entities/HdfsEntity.java]: If you've already have the ID, why do you need to generate the id?

File: model/src/test/java/com/cloudera/nav/sdk/model/entities/EntityTest.java (hunk 2 above)
Comment (changhiskhan) [reply model/src/test/java/com/cloudera/nav/sdk/model/entities/EntityTest.java]: so which unit tests are now testing entity.validateEntity() on each type we've defined?

Comment (changhiskhan) [reply model/src/test/java/com/cloudera/nav/sdk/model/entities/EntityTest.java]: Ah ok, i see the new unit tests. thanks

File: client/src/main/java/com/cloudera/nav/sdk/client/writer/MetadataWriter.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (changhiskhan) [reply client/src/main/java/com/cloudera/nav/sdk/client/writer/MetadataWriter.java]: Sorry, I wasn't 100% clear. I meant to type Entity X in both cases (a == x). Can you see how utterly confusing it's going to get for users in these very common situations?

File: model/src/main/java/com/cloudera/nav/sdk/model/relations/Relation.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (changhiskhan) [reply model/src/main/java/com/cloudera/nav/sdk/model/relations/Relation.java]: ok per our convo outside of this,  duplicated relations problem still need to be discussed with Vikas

File: client/src/main/java/com/cloudera/nav/sdk/client/writer/MetadataWriter.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (aadarshjajodia) [reply client/src/main/java/com/cloudera/nav/sdk/client/writer/MetadataWriter.java]: ok. in CustomMetadata it is clear that we are not creating any relations and hence the user would not set the endPoint flag.In CustomLineage the user knows the HdfsEntity is going to be part of a relation and hence the user sets the end point. We just need to document what setting the flag means and how to use it which should make it clear to the user. Its similar to the previous version where the user had to create an EndPointProxy in case of a CustomLineage and create an HDfsEntity in case of CustomMetadata.

[the remark this answers is not in the document]
Comment (changhiskhan) [reply client/src/main/java/com/cloudera/nav/sdk/client/writer/MetadataWriter.java]: It's not similar at all because now you're using Entity in both cases so you wouldn't expect to have to create separate instances for the same thing when using it in different places. The errors this is going to create will be utterly impossible to debug. Not sure how else I can explain it to you the evidence is abundantly clear. Don't use this pattern.

File: model/src/main/java/com/cloudera/nav/sdk/model/entities/HdfsEntity.java (hunk 34 above)
Comment (aadarshjajodia) [reply model/src/main/java/com/cloudera/nav/sdk/model/entities/HdfsEntity.java]: The user might not know the id, in which the case the user passes in the filesystem path and the source id and the server knows how to generate the id for the entity.

File: client/src/test/java/com/cloudera/nav/sdk/client/CustomOperation.java [hunk 35]
```diff
@@ -55,13 +58,10 @@ public Entity getPigOperation() {
...
-  public String getPigOperationEntityId() {
-    return pigOperation.getIdentity();
-  }
-
-  public void setPigOperationId(String pigOperationId) {
-    this.pigOperation = new EndPointProxy(pigOperationId,
-        SourceType.PIG, EntityType.OPERATION);
+  public void setPigOperation(Entity pigOperation) {
+    this.pigOperation = new EndPointProxy(
+        pigOperation.getIdAttrsMap(),
```
Comment (changhiskhan) [inline client/src/test/java/com/cloudera/nav/sdk/client/CustomOperation.java]: Does getIdAttrsMap retrieve only the id if it already exists?

**Comment by changhiskhan** [comment]
Just had that one minor comment. Outside of that I think we're probably OK for this change.
Important things:
- Looks like the unit tests are passing so that's good
- Please run all the examples
- Make sure there's clear communication about version compatibility (seek feedback from our partners please).

**Comment by changhiskhan** [comment]
Also please note that we should *NOT* use the big green "Merge pull request" button as that breaks linear commit history.
